libraries/chain/backing_store/tests/benchmark_kv.cpp

```diff
@@ -0,0 +1,337 @@
+      << "   -f=FILENAME, file storing keys one per line \n"
+      << "   -o=OPERATION, operation to be benchmarked: get, get_data, set, it_next, it_key_value, Default " << default_operation << std::endl
+      << "   -c, use Chainbase. Without this option: use RocksDb\n"
+      << "   -h, print this help\n"
+      << "   -v=NUMBER, value size. Default " << default_value_size << std::endl
+      << std::endl;
+}
+
+// As this program is only for internal use,
+// we don't use sophisticated library to parse command line
+void parse_cmd_line(int argc, char* argv[], cmd_args& args) {
+   for (auto i = 1; i < argc; ++i) {
+      std::string arg = argv[i];
+      if (arg == "-h") {
+         print_usage(argv[0]);
+         exit(0);
+      } else if (arg == "-f") {
+         if (i + 1 < argc) { 
+            args.key_file = argv[++i];
+         } else {
+            std::cerr << "-f option needs key file name" << std::endl;
+            print_usage(argv[0]);
+            exit(1);
+         }  
+      } else if (arg == "-o") {
+         if (i + 1 < argc) { 
+            args.operation = argv[++i];
+         } else {
+            std::cerr << "-o option needs operation (get, get_data, set)" << std::endl;
+            print_usage(argv[0]);
+            exit(1);
+         }  
+      } else if (arg == "-c") {  // use chainbase
+         args.is_rocksdb = false;
+      } else {
+         std::cout << "Unknown option: " << arg << std::endl;
+         print_usage(argv[0]);
+         exit(1);
+      }
+   }
+}
+
+std::vector<std::string> get_work_set_keys(const cmd_args& args) {
+   std::ifstream key_file(args.key_file);
+   if (!key_file.is_open()) {
+      std::cout << "Failed to open key file " << args.key_file << std::endl;
+      exit(2);
+   }
+
+   std::string key;
+   auto i = 0;
+   std::unordered_set<std::string> unordered_keys;
+   while (i < work_set_size && getline(key_file, key)) {
+      unordered_keys.insert(key);  // randomize
+      ++i;
+   }
+   key_file.close();
+
+   std::vector<std::string> keys;
+   for (auto& k: unordered_keys) {
+      keys.push_back(k);
+   }
+   return keys;
+}
+
+void create_key_values(cmd_args& args, const std::unique_ptr<kv_context>& kv_context_ptr) {
+   std::ifstream key_file(args.key_file);
+
+   if (!key_file.is_open()) {
+      std::cout << "Failed to open key file " << args.key_file << std::endl;
+      exit(2);
+   }
+
+   std::string value(args.value_size, 'a');
+   std::string key;
+   while (getline(key_file, key)) {
+      kv_context_ptr->kv_set(contract, key.c_str(), key.size(), value.c_str(), value.size(), payer);
+      ++args.num_keys;
+   }
+
+   key_file.close();
+}
+
+void benchmark_get(cmd_args& args, const std::unique_ptr<kv_context> kv_context_ptr, std::vector<std::string> work_set, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
+   start = std::chrono::steady_clock::now();
+   getrusage(RUSAGE_SELF, &usage_start);
```

Comment:
Is this something we want included in the performance metric?  What are we measuring in these methods?  I would think that we would want to move the measurement to only clocking the time it takes to return from `kv_get` which means the we would be counting only within the inner most loop.

Similar question in regards to the other benchmarking functions.

Reply:
getrusage returns more detailed information about performance. Bart used it in previous benchmarking activities and suggested use it here too.

Reply:
Did Bart recommend using both in this fashion?  It seems a bit counterintuitive.  You will be capturing the overhead of two `getrusage` calls within your timer.  In addition you appear to only output the time in `microseconds` which rusage already gives you.  It seems that you can get by with just the rusage and add the user time and kernel time together.

Reply:
Will only use getrusage.